Add checked-in build definitions

naamunds opened the pull request:
These build definitions will be used by pipebuild to build the product.

The signing now uses MicroBuild because the builds are being moved to
the DevDiv VSTS instance, and the Microsoft Code Signing system which
was used in MSEng is not supported there.

MicroBuild does not support the .NET Core MSBuild, so the full-framework
MSBuild is used to build the signing targets.  However, the new build
machines do not have MSBuild 15, so they use 14. This means that
projects which use the new MSBuild 15 syntax fail to be imported.
They're not needed for the signing, so their importing is skipped by
using a new SkipMsBuild15Projects property.

Thanks to @crummel for doing most of the work on the signing logic.

/cc @piotrpMSFT @livarcocc @chcosta

piotrpMSFT commented:
🥇 😍

piotrpMSFT commented on build.proj:
```diff
@@ -80,8 +80,9 @@
+  <Import Project="build/Microsoft.DotNet.Cli.Signing.targets" />
   <Import Project="build/Microsoft.DotNet.Cli.Package.targets" />
-  <Import Project="build/Microsoft.DotNet.Cli.Test.targets" />
+  <Import Project="build/Microsoft.DotNet.Cli.Test.targets" Condition="'$(SkipMsBuild15Projects)' != 'True'" />
```
Can you share the purpose of this condition, @naamunds ? The property name is not immediately intuitive to me.

naamunds replied:
MicroBuild does not support the .NET Core MSBuild, so the full-framework MSBuild is used to build the signing targets. However, the new build machines do not have MSBuild 15, so they use 14. This means that projects which use the new MSBuild 15 syntax fail to be imported.
They're not needed for the signing, so their importing is skipped by using a new `SkipMsBuild15Projects` property which is set to true in the signing build steps.

naamunds replied:
I can look into the possibility of getting MSBuild 15 on the build agents and then remove this code at that point, but I wanted to get this working now.

piotrpMSFT commented on build/Microsoft.DotNet.Cli.Prepare.targets:
```diff
@@ -2,8 +2,8 @@
-  <Import Project="Microsoft.DotNet.Cli.BundledSdks.props" />
-  <Import Project="Microsoft.DotNet.Cli.BundledTemplates.props" />
+  <Import Project="Microsoft.DotNet.Cli.BundledSdks.props" Condition="'$(SkipMsBuild15Projects)' != 'True'" />
+  <Import Project="Microsoft.DotNet.Cli.BundledTemplates.props" Condition="'$(SkipMsBuild15Projects)' != 'True'" />
```
Same here.

piotrpMSFT commented on build/Microsoft.DotNet.Cli.Prepare.targets:
```diff
@@ -68,7 +68,7 @@
     </GenerateBuildVersionInfo>
 
-    <GenerateNuGetPackagesArchiveVersion RepoRoot="$(RepoRoot)">
+    <GenerateNuGetPackagesArchiveVersion RepoRoot="$(RepoRoot)" Condition="'$(SkipMsBuild15Projects)' != 'True'">
```
and here

piotrpMSFT commented on build/Microsoft.DotNet.Cli.Signing.targets:
```diff
@@ -0,0 +1,199 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
Why ToolsVersion14? is this not running on VS 2017?

naamunds replied:
MSBuild 15 is not installed on the build agents; see the above comment thread.

piotrpMSFT commented on build/Microsoft.DotNet.Cli.Signing.targets:
```diff
@@ -0,0 +1,199 @@
+  </Target>
+
+  <Import Project="$(RepoRoot)\build_tools\MicroBuild.Core.props" />
+  <Import Project="$(RepoRoot)\build_tools\MicroBuild.Core.targets" />
```
Wohoo! a use case of build_tools :D

eerhardt replied:
This pattern is a bit odd to me.  Usually `.props` files are imported at the top of a project file, and `.targets` files are imported at the bottom.

This gives you the advantage of being able to override any property set in a `.props` file, and for the `.targets` file to be able to read properties set in your project.

naamunds replied:
Changed as suggested.

piotrpMSFT commented on build/Microsoft.DotNet.Cli.Signing.targets:
```diff
@@ -0,0 +1,199 @@
+  <Import Project="$(RepoRoot)\build_tools\MicroBuild.Core.targets" />
+
+  <Target Name="PostCompileSign" DependsOnTargets="GetPostCompileSignFiles">
+    <CallTarget Targets="SignFiles" />
```
Why do we need to use CallTarget? It's a pattern we've been trying to avoid in CLI due to the nuances [properties defined in calling targets not being available, etc]. If necessary is fine, but why not have PostCompileSign depend on SignFiles depend on GetPostCompileSignFiles?

naamunds replied:
I've changed it so that `SignFiles` is also in the `DependsOnTargets`, e.g. `DependsOnTargets="GetPostCompileSignFiles;SignFiles"`

piotrpMSFT commented on build/Microsoft.DotNet.Cli.Signing.targets:
```diff
@@ -0,0 +1,199 @@
+    <ItemGroup>
+      <!-- External files -->
+      <FilesToSign Include="$(OutputDirectory)sdk/**/Newtonsoft.Json.dll">
+        <Authenticode>$(ExternalCertificateId)</Authenticode>
```
Are we able to use "Metadata as attributes" here? Or is an older MSBuild expected to run these targets?

naamunds replied:
It looks like the "metadata as attributes" feature was supported starting with MSBuild 15, and currently the build agents don't have MSBuild 15.

piotrpMSFT commented on build/Microsoft.DotNet.Cli.Signing.targets:
```diff
@@ -0,0 +1,199 @@
+      <FilesToSign Include="$(OutputDirectory)sdk/**/Newtonsoft.Json.dll">
+        <Authenticode>$(ExternalCertificateId)</Authenticode>
+      </FilesToSign>
+      <FilesToSign Include="$(OutputDirectory)shared/**/libuv.dll">
```
This is pretty verbose. Can we have one "FilesToSign" per CertificateId and `;` separate all of these paths in a single `Include`?

naamunds replied:
Changed as suggested.

piotrpMSFT commented on build/Microsoft.DotNet.Cli.Signing.targets:
```diff
@@ -0,0 +1,199 @@
+  </Target>
+
+  <Target Name="SignNuPkgContents" DependsOnTargets="GetSignNuPkgContentsFiles">
+    <CallTarget Targets="SignFiles" />
```
Same CallTarget comment as above.

piotrpMSFT commented on build/Microsoft.DotNet.Cli.Signing.targets:
```diff
@@ -0,0 +1,199 @@
+    </PropertyGroup>
+    <ItemGroup>
+      <!-- NuPkg contents -->
+      <FilesToSign Include="$(RepoRoot)/src/Microsoft.DotNet.Cli.Utils/bin/**/*.dll">
```
Same `;` delimiter comment as above.

piotrpMSFT commented on build/Microsoft.DotNet.Cli.Signing.targets:
```diff
@@ -0,0 +1,199 @@
+  </Target>
+
+  <Target Name="SignSharedFrameworkCrossgenedAssemblies" DependsOnTargets="GetSignSharedFrameworkCrossgenedAssembliesFiles">
+    <CallTarget Targets="SignFiles" />
```
same

piotrpMSFT commented on build/Microsoft.DotNet.Cli.Signing.targets:
```diff
@@ -0,0 +1,199 @@
+      <OutDir>$(OutputDirectory)shared</OutDir>
+    </PropertyGroup>
+    <ItemGroup>
+      <FilesToSign Include="$(OutputDirectory)shared/**/Microsoft.*.dll">
```
Same

piotrpMSFT commented on build/Microsoft.DotNet.Cli.Signing.targets:
```diff
@@ -0,0 +1,199 @@
+    <CallTarget Targets="SignFiles" />
+  </Target>
+
+  <Target Name="GetSignSharedFrameworkCrossgenedAssembliesFiles" DependsOnTargets="Init">
```
This is a good time to think through this. Should CLI still be signing SharedFramework assemblies? Aren't those signed & crossgen'd in the Shared Framework packages already?

naamunds replied:
I've confirmed that these assemblies are already signed, so I'll remove that step from the build.

piotrpMSFT commented on build/Microsoft.DotNet.Cli.Signing.targets:
```diff
@@ -0,0 +1,199 @@
+  </Target>
+
+  <Target Name="SignMsiAndCab" DependsOnTargets="GetSignMsiAndCabFiles">
+    <CallTarget Targets="SignFiles" />
```
Same

piotrpMSFT commented on build/Microsoft.DotNet.Cli.Signing.targets:
```diff
@@ -0,0 +1,199 @@
+      <OutDir>$(PackagesDirectory)</OutDir>
+    </PropertyGroup>
+    <ItemGroup>
+      <FilesToSign Include="$(PackagesDirectory)/**/*.msi">
```
Same

piotrpMSFT commented on build/Microsoft.DotNet.Cli.Signing.targets:
```diff
@@ -0,0 +1,199 @@
+  </Target>
+
+  <Target Name="SignEngine" DependsOnTargets="GetSignEngineFiles">
+    <CallTarget Targets="SignFiles" />
```
Same

piotrpMSFT commented on build/Microsoft.DotNet.Cli.Signing.targets:
```diff
@@ -0,0 +1,199 @@
+  </Target>
+
+  <Target Name="SignCliBundle" DependsOnTargets="GetSignCliBundleFiles">
+    <CallTarget Targets="SignFiles" />
```
Same

piotrpMSFT commented on build/Microsoft.DotNet.Cli.Signing.targets:
```diff
@@ -0,0 +1,199 @@
+    <ParameterGroup>
+      <SearchPath ParameterType="System.String" Required="true" />
+    </ParameterGroup>
+    <Task>
```
This is the first inline task in CLI repo. Can we instead put this in dotnet-cli-build.csproj?

naamunds replied:
I just tried that, but `AppDomain` isn't going to be added to .NET Core until 2.0 (see https://apisof.net/catalog/System.AppDomain), so the build of dotnet-cli-build.csproj fails. I think for this reason we'll have to leave it as an inline task, where it can use the full-framework .NET.

piotrpMSFT commented on buildpipeline/Dotnet-CLI-Ubuntu16.04-x64.json:
```diff
@@ -0,0 +1,180 @@
+    "state": "wellFormed",
+    "revision": 418097459
+  }
+}
```
Please make sure we have all of these trailing newline issues resolved.

naamunds replied:
Fixed.

piotrpMSFT commented on buildpipeline/DotNet-CLI-CentOS-x64.json:
```diff
@@ -0,0 +1,172 @@
+{
```
This is hopeful: Can we collapse these down to a single file for common file? having these files be disjoint will make it difficult to understand where our builds differ...

naamunds replied:
Yes, that's the long-term plan. I wanted to get this in quickly by doing the work iteratively; I'll log an issue for consolidating the build definitions.

eerhardt commented on build/Microsoft.DotNet.Cli.Signing.targets:
```diff
@@ -0,0 +1,199 @@
+
+  <Target Name="GetSignNuPkgContentsFiles" DependsOnTargets="Init">
+    <PropertyGroup>
+      <OutDir>$(RepoRoot)/src/Microsoft.DotNet.Cli.Utils</OutDir>
```
seems like an odd "outdir".  Is this used and correct?

naamunds replied:
Yes; this property refers to the directory in which the files to sign are located. MicroBuild only signs files that are in a subdirectory of either the `OutDir` or `IntermediateOutputPath`. For this task, the files to sign are in the `$(RepoRoot)/src/Microsoft.DotNet.Cli.Utils` directory.

eerhardt commented on build/Microsoft.DotNet.Cli.Signing.targets:
```diff
@@ -0,0 +1,199 @@
+      <FilesToSign Include="$(RepoRoot)/src/Microsoft.DotNet.Cli.Utils/bin/**/*.exe">
+        <Authenticode>$(InternalCertificateId)</Authenticode>
+      </FilesToSign>
+      <FilesToSign Include="$(RepoRoot)/src/Microsoft.DotNet.Cli.Utils/obj/**/*.dll">
```
Why are we signing files in the `obj` folder?  Seems like only files in the `bin` folder are necessary.

eerhardt replied:
Also, won't this sign a bunch of referenced assemblies as well?  I don't think we need them signed - only what gets put in the nupkg.

naamunds replied:
I'm fine with removing the obj files and only signing `$(RepoRoot)/src/Microsoft.DotNet.Cli.Utils/bin/**/Microsoft.DotNet.Cli.Utils.dll` here.  I tested out that change, and it results in every file in the nupkg being signed. Do we know that Microsoft.DotNet.Cli.Utils.dll will always be the only file in the nupkg, though?

naamunds commented:
In addition to addressing the comments here, I've updated the Windows x64 build definition to use a PowerShell script for publishing the NuGet packages to the dotnet-cli MyGet feed.  This is because the NuGet Server Endpoint used by the NuGet Publisher build task in MSEng does not exist in DevDiv, and because the PS script as easier to maintain.

naamunds commented:
@piotrpMSFT I've moved the signing logic to a standalone project due to it needing to be built with the .NET desktop MSBuild, as we discussed offline.  This decoupling gives a couple simplifying benefits:

- We no longer need to use `Condition="'$(SkipMsBuild15Projects)' != 'True'"`
- The HandleAssemblyResolve inline task is no longer necessary; it was needed for calculating the RID as part build.proj.

When we have a common props file with the properties at https://github.com/naamunds/cli/blob/94648355485c08c2f3776eb128ec3b28c8be2554/build/Microsoft.DotNet.Cli.Signing.proj#L17-L20, the code can be refactored to use them.

piotrpMSFT commented on build/Microsoft.DotNet.Cli.Signing.proj:
```diff
@@ -0,0 +1,117 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project InitialTargets="SetSigningProperties" DefaultTargets="SignFiles" ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <RepoRoot>$(MSBuildThisFileDirectory)/..</RepoRoot>
```
Consider using CLI's `dir.props`. It's great for properties that we already define statically like `<RepoRoot>`

naamunds replied:
Good idea.

naamunds commented:
@piotrpMSFT @livarcocc I've added a commit porting over https://github.com/dotnet/core-setup/pull/1406.  This issue would impact CLI also if not fixed; I saw it in some of my test RHEL builds.  As I mentioned there, I see this as a short-term fix.  Let me know if you have any questions or concerns.

naamunds commented:
I've added https://github.com/dotnet/cli/pull/5489/commits/4322dd5601e4b7ddc32398502b3e7886524106a2 for temporarily using sandboxes for publishing. @piotrpMSFT @livarcocc PTAL.

/cc @dagood

piotrpMSFT commented:
@dotnet-bot test Windows_NT x64 Release Build please.

piotrpMSFT commented:
@dotnet-bot test Windows_NT x64 Release Build please.

piotrpMSFT commented:
@MattGertz merging the infra PR we approved last night.
